File: leetcode53/leetcode53.py
# class Solution:
#     def maxSubArray(self, nums) -> int:
#         dp = [0] * len(nums)
#         dp[0] = nums[0]
#         res = nums[0]
#         for i in range(1, len(nums)):
#             dp[i] = max(dp[i - 1] + nums[i], nums[i])
#             if dp[i] > res:
#                 res = dp[i]
#         return res

# sol = Solution()
# print(sol.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))

class Solution:
    def maxSubArray(self, nums) -> int:
        dp = [0] * len(nums)
        dp[0] = nums[0]
        res = nums[0]
        for i in range(1, len(nums)):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
            res = max(res, dp[i])
        return res

sol = Solution()
print(sol.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))

File: leetcode53/leetcode53.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() == 0) return 0;
        vector<int> dp(nums.size());
        dp[0] = nums[0];
        int res = dp[0];
        for (int i = 1; i < nums.size(); i++){
            dp[i] = max(dp[i - 1] + nums[i], nums[i]);
            if(dp[i] > res) res = dp[i];
        }
        return res;
    }
};

int main(){
    Solution sol;
    vector<int> nums = {-2,1,-3,4,-1,2,1,-5,4};
    cout << sol.maxSubArray(nums) << endl;
    return 0;
}
